src/inference/Inference.jl

```diff
@@ -77,24 +84,368 @@ function mh_accept(H::T, H_new::T, log_proposal_ratio::T) where {T<:Real}
+                name = string(ind2sub(size(v), i))
+                name = replace(name, "(" => "[");
+                name = replace(name, ",)" => "]");
+                name = replace(name, ")" => "]");
+                name = k * name
+                isa(v[i], Nothing) && println(v, i, v[i])
+                push!(value, v[i])
+                push!(names, name)
+            elseif isa(v[i], AbstractArray)
+                name = k * string(ind2sub(size(v), i))
+                flatten(names, value, name, v[i])
+            else
+                error("Unknown var type: typeof($v[i])=$(typeof(v[i]))")
+            end
+        end
+    else
+        error("Unknown var type: typeof($v)=$(typeof(v))")
+    end
+    return
+end
+
+ind2sub(v, i) = Tuple(CartesianIndices(v)[i])
+
+function get_transition_extras(ts::Vector{T}) where T<:AbstractTransition
+    # Get the extra field names from the sampler state type.
+    # This handles things like :lp or :weight.
+    extra_params = additional_parameters(T)
+
+    # Get the values of the extra parameters.
+    local extra_names
+    all_vals = []
+    for t in ts
+        extra_names = String[]
+        vals = []
+        for p in extra_params
+            prop = getproperty(t, p)
+            if prop isa NamedTuple
+                for (k, v) in pairs(prop)
+                    push!(extra_names, string(k))
+                    push!(vals, v)
+                end
+            else
+                push!(extra_names, string(p))
+                push!(vals, prop)
+            end
+        end
+        push!(all_vals, vals)
+    end
+
+    return extra_names, vcat(all_vals)
+end
+
+# Default Chains constructor.
+function Chains(
+    rng::AbstractRNG,
+    model::ModelType,
+    spl::Sampler,
+    N::Integer,
+    ts::Vector{T};
+    discard_adapt::Bool=true,
+    save_state=false,
+    kwargs...
+) where {ModelType<:Sampleable, T<:AbstractTransition}
+    # Check if we have adaptation samples.
+    if discard_adapt && :n_adapts in fieldnames(typeof(spl.alg))
+        ts = ts[(spl.alg.n_adapts+1):end]
+    end
+
+    # Convert transitions to array format.
+    # Also retrieve the variable names.
+    nms, vals = _params_to_array(ts, spl)
+
+    # Get the values of the extra parameters in each Transition struct.
+    extra_params, extra_values = get_transition_extras(ts)
+
+    # Extract names & construct param array.
+    nms = string.(vcat(nms..., string.(extra_params)...))
+    parray = vcat([hcat(vals[i]..., extra_values[i]...) for i in 1:length(ts)]...)
+
+    # If the state field has average_logevidence or final_logevidence, grab that.
+    le = missing
+    if :average_logevidence in fieldnames(typeof(spl.state))
+        le = getproperty(spl.state, :average_logevidence)
+    elseif :final_logevidence in fieldnames(typeof(spl.state))
+        le = getproperty(spl.state, :final_logevidence)
+    end
+
+    # Check whether to invlink! the varinfo
+    if islinked(spl.state.vi, spl)
```

Comment:
Shouldn't this be part of the `sample_end!`?

Reply:
I've only added it here in case the state needs to be saved in the chain. At current most samplers don't actually need to call `invlink!` at the end of the sample, so it's ignored. This is one of those things I would rather just have `VarInfo` do for me -- all the inference should be in the transformed space, and anything that the user sees is in the constrained space.

Reply:
We can discuss this in a separate issue.